server: rework cluster version initialization

tbg opened the pull request:
A central invariant around cluster versions is that when an update
arrives, we need to persist it before exposing it through the version
setting. This was not true during server start time, as described in:

https://github.com/cockroachdb/cockroach/pull/47235#issuecomment-611419920

In short, we were registering the callback to persist to the engines
*after* Gossip had already connected, opening up a window during which
a cluster version bump simply would not be persisted.

In the acceptance/version-upgrade test, this would manifest as nodes
refusing to start because their binary version had proceeded to far
beyond the persisted version. At the time of writing and before this
commit, this would happen in perhaps 1-10% of local runs on a linux
machine (rarely on OSX).

The buggy code was originally written when the startup sequence was a
lot more intricate and we were, as I recall, under pressure to deliver.
Now, with recent refactors around the `initServer`, we're in a
really good place to solve this while also simplifying the whole
story. In this commit,

- persist the cluster version on *all* engines, not just the initialized
  ones; this completely decouples the timing of when the engines get
  initialized from when we can set up the persistence callback and
  makes everything *much simpler*.
- stop opportunistically backfilling the cluster version. It is now
  done once at the beginning, in the right place, without FUD later
  on.
- remove the cluster version persistence from engine initialization.
  Anyone initializing an engine must put a cluster version on it
  first. In a running server, this happens during initServer creation
  time, way before there are any moving parts in the system. In tests,
  extra writes were added as needed.
- set up the callback with Gossip before starting Gossip, and make sure
  (via an assertion) that this property does not rot.
  By setting up the callback before Gossip starts, we make sure there
  isn't a period during which Gossip receives an update but doesn't have
  the callback yet.
- as a result of all of the above, take all knowledge of cluster version
  init away from `*Node` and `*Stores`.

As a last note, we are planning to stop using Gossip for this version
business in 20.1, though this too will be facilitated by this change.

Release note (bug fix): Avoid a condition during rapid version upgrades
where a node would refuse to start, claiming "[a store is] too old for
running version". Before this bug fix, the workaround is to decommission
the node, delete the store directory, and re-add it to the cluster as
a new node.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47358)
<!-- Reviewable:end -->

tbg committed:
server: avoid log spam if node down

If a cluster is fully upgraded but a node is down, nodes that start will
not manage to exit the cluster version auto upgrade worker because the
down node would take precedence over the realization that no upgrade
was necessary.

Release note: None

irfansharif reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> everything looks much simpler now.

> Reviewed 1 of 1 files at r1, 17 of 17 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-:-M4pNp181CRXq7oGw-pq:b7owtoh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/clusterversion/setting.go, line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pMwyRDp4Ovpl6f7D0:-M4pMwyRDp4Ovpl6f7D1:btk063h) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/clusterversion/setting.go#L97)):*
> ```Go
> 		//
> 		// It's also used in production code during bootstrap, where the version
> 		// is first initialized to MinSupportedVersion and then re-initialized
> ```

s/MinSupportedVersion/BinaryMinSupportedVersion to be consistent with the signatures throughout?
___
*[pkg/kv/kvserver/replica_test.go, line 243 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pH8vZ1CGmZt5T-9xF:-M4pH8vZ1CGmZt5T-9xG:b-5j8k7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/replica_test.go#L243)):*
> ```Go
> 		cfg.DB = kv.NewDB(cfg.AmbientCtx, factory, cfg.Clock)
> 
> 		require.NoError(t, WriteClusterVersion(ctx, tc.engine, clusterversion.ClusterVersion{Version: bootstrapVersion}))
> ```

Use `cv` instead?
___
*[pkg/kv/kvserver/store.go, line 2022 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pF20j0wnESIAcZRas:-M4pF20j0wnESIAcZRat:b-jv3m8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/store.go#L2022)):*
> ```Go
> }
> 
> // checkCanInitializeEngine ensures that the engine is empty except for a a
> ```

Double "a".
___
*[pkg/kv/kvserver/store.go, line 2031 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pFF8GCjzoYlY4UO8X:-M4pFF8GCjzoYlY4UO8Y:b-rbf610) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/store.go#L2031)):*
> ```Go
> 	// See if this is an already-bootstrapped store.
> 	ident, err := ReadStoreIdent(ctx, eng)
> 	if err != nil {
> ```

[nit]: Maybe check for `if err == nil { ... }` and then separately `if !errors.Is[...] { ... }`, to make the structure simpler.
___
*[pkg/kv/kvserver/stores.go, line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pFhM44FRRpV3PF3Ii:-M4pFhM44FRRpV3PF3Ij:b-z13dzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/stores.go#L280)):*
> ```Go
> // engine, falling back to the zero value.
> func ReadVersionFromEngineOrZero(
> 	ctx context.Context, e storage.Reader,
> ```

[nit] Rename "e" to "r" or "reader".
___
*[pkg/kv/kvserver/stores.go, line 315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pGALhC7NjSLuqckW-:-M4pGALhC7NjSLuqckW0:bag7c9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/stores.go#L315)):*
> ```Go
> // 	binaryVersion: The version of this binary. An error is returned if
> // 	  any engine has a higher version, as this would indicate that this node
> //    has previously acked the higher cluster version but is now running an
> ```

Fix indentation here.
___
*[pkg/kv/kvserver/stores.go, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pGQEW27S9aiqP57w4:-M4pGQEW27S9aiqP57w5:bsxjct6) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/stores.go#L342)):*
> ```Go
> 	// then minStoreVersion don't change any more).
> 	for _, eng := range engines {
> 		eng := eng.(storage.Reader) // we're read only
> ```

[nit] "We're read only."
___
*[pkg/kv/kvserver/stores_test.go, line 474 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pHpBmAub-BDZkJbJO:-M4pHpBmAub-BDZkJbJP:b-k6msy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/stores_test.go#L474)):*
> ```Go
> 
> 	type testCase struct {
> 		binV, minV roachpb.Version // min supported version and binary version
> ```

Reorder versions in comment.
___
*[pkg/server/init.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pLamGE1MoZlIOmpF3:-M4pLamGE1MoZlIOmpF4:b-t9qht) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/init.go#L175)):*
> ```Go
> 		// Bootstrap() did its job. At this point, we know that the cluster version
> 		// will be bootstrapVersion (=state.clusterVersion.Version), but the version
> 		// setting does not know yet (it was initialized as the BinaryMinSupported
> ```

nit: don't break up "BinaryMinSupportedVersion" into separate lines, would make it harder to grep for.
___
*[pkg/server/init.go, line 269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pNGmrCXxMAln9TDmm:-M4pNGmrCXxMAln9TDmn:b4xmium) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/init.go#L269)):*
> ```Go
> 
> // DiskClusterVersion returns the cluster version synthesized from disk. This
> // is always non-zero since it falls back to the MinSupportedVersion.
> ```

s/MinSupportedVersion/BinaryMinSupportedVersion, same comment as elsewhere.
___
*[pkg/server/node.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pJMauBa84vvq8-5sy:-M4pJMauBa84vvq8-5sz:be0wejb) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/node.go#L225)):*
> ```Go
> 		}
> 
> 		if bootstrapVersion == (clusterversion.ClusterVersion{}) {
> ```

Would `i == 0` work better here?
___
*[pkg/server/server.go, line 1519 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pJm2KDbNDbgKuKl23:-M4pJm2KDbNDbgKuKl24:b71nk75) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/server.go#L1519)):*
> ```Go
> 			defer mu.Unlock()
> 			v := s.cfg.Settings.Version
> 			prevCV, err := kvserver.SynthesizeClusterVersionFromEngines(ctx, s.engines, v.BinaryVersion(), v.BinaryMinSupportedVersion())
> ```

[nit] Super long line.
___
*[pkg/server/server.go, line 1538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pKJbx5gV6Mi5BP_D0:-M4pKJbx5gV6Mi5BP_D1:b9dezdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/server.go#L1538)):*
> ```Go
> 
> 		diskClusterVersion := initServer.DiskClusterVersion()
> 		// The version setting loaded from disk (in setupInitServer above) is
> ```

There's no `setupInitServer` above, can do without the phrase anyway.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Thanks Irfan!

bors r=irfansharif

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-:-M4skuvbBNRQBK_ngkJt:b-etty07)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @irfansharif)

___
*[pkg/clusterversion/setting.go, line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pMwyRDp4Ovpl6f7D0:-M4shnE0-bcf1bvjSTOi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/clusterversion/setting.go#L97)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/MinSupportedVersion/BinaryMinSupportedVersion to be consistent with the signatures throughout?
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_test.go, line 243 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pH8vZ1CGmZt5T-9xF:-M4shteW1qdz0rMG1s65:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/replica_test.go#L243)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Use `cv` instead?
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store.go, line 2022 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pF20j0wnESIAcZRas:-M4shwYz-8tH0Us_oUVK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/store.go#L2022)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Double "a".
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store.go, line 2031 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pFF8GCjzoYlY4UO8X:-M4silRS7Dr-N2vUDQ_f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/store.go#L2031)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit]: Maybe check for `if err == nil { ... }` and then separately `if !errors.Is[...] { ... }`, to make the structure simpler.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/stores.go, line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pFhM44FRRpV3PF3Ii:-M4siqOn9WZ7pPTI1pb_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/stores.go#L280)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Rename "e" to "r" or "reader".
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/stores.go, line 315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pGALhC7NjSLuqckW-:-M4sj7qD9nmmQhhhUKlI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/stores.go#L315)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Fix indentation here.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/stores.go, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pGQEW27S9aiqP57w4:-M4sj9cv4ZAqNuBe_H5S:b-jef2b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/stores.go#L342)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] "We're read only."
</blockquote></details>

I can't find it explicitly any more, but you can see on https://github.com/golang/go/wiki/CodeReviewComments that inline comments don't capitalize or punctuate. Or I'm missing your point because you're not pointing this out in other places in this review?
___
*[pkg/kv/kvserver/stores_test.go, line 474 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pHpBmAub-BDZkJbJO:-M4sjubTDVCoIlMQNScP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/kv/kvserver/stores_test.go#L474)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Reorder versions in comment.
</blockquote></details>

Done.
___
*[pkg/server/init.go, line 269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pNGmrCXxMAln9TDmm:-M4sk2Do5FJCg_iKWGev:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/init.go#L269)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/MinSupportedVersion/BinaryMinSupportedVersion, same comment as elsewhere.
</blockquote></details>

Done.
___
*[pkg/server/node.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pJMauBa84vvq8-5sy:-M4sk9CDBeraycbO8yIP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/node.go#L225)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Would `i == 0` work better here?
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 1519 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pJm2KDbNDbgKuKl23:-M4skCJY8RRqSu9fpNrv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/server.go#L1519)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Super long line.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 1538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47358#-M4pKJbx5gV6Mi5BP_D0:-M4skHee1BVfcfPOE7_y:bxva74c) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bf85654b1cb5b8513e8685f5ecf65247b641ac/pkg/server/server.go#L1538)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

There's no `setupInitServer` above, can do without the phrase anyway.
</blockquote></details>

There is! But it's really far away. Removed.


<!-- Sent from Reviewable.io -->

tbg committed:
server: rework cluster version initialization

A central invariant around cluster versions is that when an update
arrives, we need to persist it before exposing it through the version
setting. This was not true during server start time, as described in:

https://github.com/cockroachdb/cockroach/pull/47235#issuecomment-611419920

In short, we were registering the callback to persist to the engines
*after* Gossip had already connected, opening up a window during which
a cluster version bump simply would not be persisted.

In the acceptance/version-upgrade test, this would manifest as nodes
refusing to start because their binary version had proceeded to far
beyond the persisted version. At the time of writing and before this
commit, this would happen in perhaps 1-10% of local runs on a linux
machine (rarely on OSX).

The buggy code was originally written when the startup sequence was a
lot more intricate and we were, as I recall, under pressure to deliver.
Now, with recent refactors around the `initServer`, we're in a
really good place to solve this while also simplifying the whole
story. In this commit,

- persist the cluster version on *all* engines, not just the initialized
  ones; this completely decouples the timing of when the engines get
  initialized from when we can set up the persistence callback and
  makes everything *much simpler*.
- stop opportunistically backfilling the cluster version. It is now
  done once at the beginning, in the right place, without FUD later
  on.
- remove the cluster version persistence from engine initialization.
  Anyone initializing an engine must put a cluster version on it
  first. In a running server, this happens during initServer creation
  time, way before there are any moving parts in the system. In tests,
  extra writes were added as needed.
- set up the callback with Gossip before starting Gossip, and make sure
  (via an assertion) that this property does not rot.
  By setting up the callback before Gossip starts, we make sure there
  isn't a period during which Gossip receives an update but doesn't have
  the callback yet.
- as a result of all of the above, take all knowledge of cluster version
  init away from `*Node` and `*Stores`.

As a last note, we are planning to stop using Gossip for this version
business in 20.1, though this too will be facilitated by this change.

Release note (bug fix): Avoid a condition during rapid version upgrades
where a node would refuse to start, claiming "[a store is] too old for
running version". Before this bug fix, the workaround is to decommission
the node, delete the store directory, and re-add it to the cluster as
a new node.

tbg commented:
bors r=irfansharif

Forgot that I wanted to re-skip acceptance/version-upgrade to rebuild fixtures.
